flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -672,6 +644,64 @@ private void startTriggeringCheckpoint(CheckpointTriggerRequest request) {
         }
     }
 
+    private void triggerCheckpointRequest(
+            CheckpointTriggerRequest request, long timestamp, PendingCheckpoint checkpoint) {
+        if (checkpoint.isDisposed()) {
+            onTriggerFailure(
+                    checkpoint,
+                    new CheckpointException(
+                            CheckpointFailureReason.TRIGGER_CHECKPOINT_FAILURE,
+                            checkpoint.getFailureCause()));
+        } else {
+            triggerTasks(request, timestamp, checkpoint)
+                    .exceptionally(
+                            failure -> {
+                                onTriggerFailure(
+                                        checkpoint,
+                                        new CheckpointException(
+                                                CheckpointFailureReason.TRIGGER_CHECKPOINT_FAILURE,
+                                                failure));
+                                return null;
+                            });
+
+            coordinatorsToCheckpoint.forEach(
+                    (ctx) -> ctx.afterSourceBarrierInjection(checkpoint.getCheckpointID()));
+            // It is possible that the tasks has finished
+            // checkpointing at this point.
+            // So we need to complete this pending checkpoint.
+            if (maybeCompleteCheckpoint(checkpoint)) {
+                onTriggerSuccess();
```

Comment:
There might be one problem for `onTriggerSuccess` here  that we might first call `onTriggerSuccess` here, and later we found trigger failure and call `onTriggerFailure` in the callback. This might cause some problem like it would modify `isTriggering = false` and schedule the next trigger task between the current trigger (which composes of several separate Runnables), and it might also have some semantics conflict. Perhaps we might change to one of the followings ? 
1. Call `onTriggerSuccess()` also in the callback of trigger tasks finished. But if the trigger actually failed with timeout, then the trigger result check would be hold by rpc timeout (default 10s) and between this period we could not trigger new checkpoints. 
2. We think the checkpoint is triggered successfully after this line is executed, and if we found triggering tasks fails we only call `abortPendingCheckpoint`. If we execute the callback also in timer thread, there would be no concurrent issue since `abortPendingCheckpoint` would always submitted after the current method is done.

Reply:
Is it possible that triggering a checkpoint on a Task fails, while the global checkpoint succeeds? I did not put much thought into the race conditions here, cause I assumed that the triggering must succeed in order for the checkpoint to complete.

Reply:
I think the method name might have some misleadings in that it also return `true` if the checkpoint is not done yet, and it only returns false if it found that the checkpoint could not complete in the future~

Reply:
Ah, you're right. I'll look into it a bit more.